File: content/scala-driver/v5.5/source/includes/usage-examples/ReadCodeExamples.scala
import org.mongodb.scala._
import org.mongodb.scala.bson.Document
import org.mongodb.scala.model.Filters._
import org.mongodb.scala.model.changestream._

object ReadCodeExamples {

  def main(args: Array[String]): Unit = {
    val mongoClient = MongoClient("<connection string URI>")
    val database: MongoDatabase = mongoClient.getDatabase("<database name>")
    val collection: MongoCollection[Document] = database.getCollection("<collection name>")

    {
      // start-find-one
      val filter = equal("<field>", "<value>")

      collection.find(filter).first().subscribe((doc: Document) => println(doc.toJson()),
        (e: Throwable) => println(s"There was an error: $e"))
      // end-find-one
    }

    {
      // start-find-multiple
      val filter = equal("<field>", "<value>")

      collection.find(filter).subscribe((doc: Document) => println(doc.toJson()),
        (e: Throwable) => println(s"There was an error: $e"))
      // end-find-multiple
    }

    {
      // start-count-collection
      collection.countDocuments()
        .subscribe((count: Long) => println(s"Number of documents: $count"),
          (e: Throwable) => println(s"There was an error: $e"))
      // end-count-collection
    }

    {
      // start-count-documents
      collection.countDocuments(equal("<field>", "<value>"))
        .subscribe((count: Long) => println(s"Number of documents: $count"),
          (e: Throwable) => println(s"There was an error: $e"))
      // end-count-documents
    }

    {
      // start-estimate-count
      collection.estimatedDocumentCount()
        .subscribe((count: Long) => println(s"Estimated number of documents: $count"),
          (e: Throwable) => println(s"There was an error: $e"))
      // end-estimate-count
    }

    {
      // start-retrieve-distinct
      collection.distinct("<field>")
        .subscribe((value: String) => println(value),
          (e: Throwable) => println(s"There was an error: $e"))
      // end-retrieve-distinct
    }

    {
      // start-monitor-changes
      val changeStreamObservable = collection.watch()

      changeStreamObservable.subscribe(
        (changeEvent: ChangeStreamDocument[Document]) => {
          println(s"Received a change to the collection: ${changeEvent}")
        },
        (e: Throwable) => {
          println(s"Encountered an error: ${e.getMessage}")
        },
        () => println("Completed")
      )
      // end-monitor-changes
    }

    // Wait for the operations to complete before closing client
    Thread.sleep(1000)
    mongoClient.close()
  }
}

File: content/ruby-driver/v2.21/source/includes/usage-examples/read-code-examples.rb
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'mongo'
end

uri = '<connection string>'

Mongo::Client.new(uri) do |client|

  # start-find-one
  document = collection.find(name: '<value>').first
  puts document
  # end-find-one

  # start-find-many
  results = collection.find(founded_year: '<value>')
  # end-find-many

  # start-count-collection
  result = collection.count_documents
  puts "Number of documents: #{result}"
  # end-count-collection

  # start-count-accurate
  result = collection.count_documents('key': '<value>')
  puts "value: #{result}"
  # end-count-accurate

  # start-count-estimate
  result = collection.estimated_document_count
  puts "Estimated number of documents: #{result}"
  # end-count-estimate

  # start-distinct
  results = collection.distinct('field')
  # end-distinct

  # start-monitor-changes
  stream = collection.watch
  collection.insert_one(a: 1)
  doc = stream.first
  process(doc)
  # end-monitor-changes